### PR TITLE
Only lookup types in one interner

### PR DESCRIPTION
**Pull request by Zoxc**

### CODE REVIEW

**Comment by Zoxc** [comment]
@Mark-Simulacrum I'd like a perf run here

**Comment by Mark-Simulacrum** [comment]
Perf queued (and started).

**Comment by Zoxc** [comment]
Results: http://perf.rust-lang.org/compare.html?start=9ff8ec8fdf02b6d8ba35373943e0c6de13597697&end=0f8ec0a707d11fe3ffeccd1e4bafe9ca8922b42c&stat=instructions%3Au

File: src/librustc/ty/flags.rs [hunk 1]
```diff
@@ -250,3 +250,160 @@ impl FlagComputation {
         }
     }
 }
+
+pub(crate) fn sty_in_local_arena(st: &ty::TypeVariants) -> bool {
```
Comment (eddyb) [inline src/librustc/ty/flags.rs]: Can't you reuse `TypeVisitor` for this? Seems like a lot of code IMO.

**Comment by eddyb** [comment]
LGTM, modulo the implementation duplicating a lot of logic.

**Comment by Zoxc** [comment]
@Mark-Simulacrum I'd like another perf run to see if the fast path helps

**Comment by kennytm** [comment]
Perf link when it is available: http://perf.rust-lang.org/compare.html?start=6a87289fa4f49b6bdd62f33f69a580026223421f&end=116cf1c0ba1e36f4441db6653e871bc8d3d3f670&stat=instructions%3Au

**Comment by kennytm** [comment]
@Zoxc @eddyb  Perf result is now available.

**Comment by Zoxc** [comment]
It seems like the fast-path wasn't a win. I've removed it and also applied the transformation to the interning macro.

**Comment by michaelwoerister** [comment]
lgtm! You can still r- if you have any concerns, @eddyb.

@bors r+

**Pull request merged**